runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java

```diff
@@ -137,29 +160,37 @@ protected OutputT computeNext() {
       // collection (and iterator) is reset between each call to processElement, so the
       // collection only holds the output values for each call to processElement, rather
       // than for the whole partition (which would use too much memory).
+
+      boolean isBundleStarted = false;
+      boolean isBundleFinished = false;
       while (true) {
         if (outputIterator.hasNext()) {
           return outputIterator.next();
-        } else if (inputIterator.hasNext()) {
-          clearOutput();
+        }
+
+        if (!isBundleStarted) {
+          isBundleStarted = true;
+          // call startBundle() before beginning to process the partition.
+          doFnRunner.startBundle();
+        }
+
+        clearOutput();
+        if (inputIterator.hasNext()) {
           // grab the next element and process it.
           doFnRunner.processElement(inputIterator.next());
           outputIterator = getOutputIterator();
         } else if (timerDataIterator.hasNext()) {
-          clearOutput();
           fireTimer(timerDataIterator.next());
           outputIterator = getOutputIterator();
         } else {
           // no more input to consume, but finishBundle can produce more output
-          if (!calledFinish) {
-            clearOutput();
-            calledFinish = true;
+          if (!isBundleFinished) {
+            isBundleFinished = true;
             doFnRunner.finishBundle();
-            // teardown DoFn.
-            DoFnInvokers.invokerFor(doFn).invokeTeardown();
             outputIterator = getOutputIterator();
             continue; // try to consume outputIterator from start of loop
           }
+          DoFnInvokers.invokerFor(doFn).invokeTeardown();
```

Comment:
So I'm new to this part of the code, so please correct me if I'm wrong, but this will now invoke tear down each time it constructs a new new output iterator for a given elem off the input iterator and it seems like we aren't done with the DoFn at this point necessarily.

Reply:
Hi I am new to this code too, so I might be misinterpreting this, but what I understood is that once we are in the else we are 'outputtng' and in that case we have two scenarios we haven't finished the bundle so we do (notice the continue), or we can be iterating on an already finished bundle and then we just invoke the teardown. So this code 'evaluates' the optimistic path. Notice that the Wrapped iterator deals with invoking the teardown in case of exceptions.